core/src/main/java/com/microsoft/applicationinsights/ByteBufferPool.java

```diff
@@ -0,0 +1,29 @@
+package com.microsoft.applicationinsights;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ArrayBlockingQueue;
+
+class ByteBufferPool {
+
+    private static final int BYTE_BUFFER_SIZE = 65536;
+    private static final int MAX_RETAINED = 10;
+
+    private final Queue<ByteBuffer> queue = new ArrayBlockingQueue<>(MAX_RETAINED);
+
+    ByteBuffer remove() {
+        ByteBuffer byteBuffer = queue.poll();
+        if (byteBuffer != null) {
+            return byteBuffer;
+        }
+        return ByteBuffer.allocate(BYTE_BUFFER_SIZE);
+    }
+
+    void offer(List<ByteBuffer> byteBuffers) {
+        // TODO(trask) batch offer?
+        for (ByteBuffer byteBuffer : byteBuffers) {
+            queue.offer(byteBuffer);
```

Comment:
can you check bool return type?  when the queue is full, it doesn't throw an exception instead it returns false and byteBuffer is not added to the queue, which might lead to data lost.

Reply:
data has already been sent, so no data lost. it's ok if pool is full, the byte buffer will be GC'd in that case